### PR TITLE
batch copy or move process #2753

### PR DESCRIPTION
**Pull request by zixi0825**
## What is the purpose of the pull request

This pull request adds batch copy or move process feature backend code #2753

### CODE REVIEW

**Comment by break60** [comment]
Please solve ut

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java [hunk 1]
```diff
@@ -295,36 +295,188 @@ private String getResourceIds(ProcessData processData) {
...
+    }
+
+    /**
+     * check batch operate result
+     * @param result
+     * @param failedIdList
+     */
+    private void checkBatchOperateResult(Map<String, Object> result, List<String> failedIdList) {
+        if (!failedIdList.isEmpty()) {
+            putMsg(result, Status.MOVE_PROCESS_DEFINITION_ERROR, String.join(",", failedIdList));
```
Comment (Rubik-W) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java:477]: It is recommended the error message return project name.

**Comment by Rubik-W** [comment]
1. After confirming the move workflow, does the modification on the workflow instance list page have an impact?
2. Confirm there are no file changes on the front-end page ?

**Comment by gabrywu** [comment]
@zixi0825 I think Rubik-W has already gave some good suggestions, can you reply to him ?

**Comment by zixi0825** [comment]
> 1. After confirming the move workflow, does the modification on the workflow instance list page have an impact?
> 2. Confirm there are no file changes on the front-end page ?

to 1 : I will add the instance linkage after move workflow
to 2：The modification of the front-end will be completed by @break60

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ProcessDefinitionController.java [hunk 2]
```diff
@@ -95,27 +95,33 @@ public Result createProcessDefinition(@ApiIgnore @RequestAttribute(value = Const
...
-        logger.info("copy process definition, login user:{}, project name:{}, process definition id:{}",
-                loginUser.getUserName(), projectName, processId);
-        Map<String, Object> result = processDefinitionService.copyProcessDefinition(loginUser, projectName, processId);
-        return returnDataList(result);
+    @ApiException(COPY_OR_MOVE_PROCESS_DEFINITION_ERROR)
+    public Result copyOrMoveProcessDefinition(@ApiIgnore @RequestAttribute(value = Constants.SESSION_USER) User loginUser,
+                                              @ApiParam(name = "projectName", value = "PROJECT_NAME", required = true) @PathVariable String projectName,
+                                              @RequestParam(value = "processDefinitionIds", required = true) String processDefinitionIds,
+                                              @RequestParam(value = "targetProjectName",required = true) String targetProjectName,
+                                              @RequestParam(value = "isCopy", required = true) boolean isCopy)  {
```
[the remark this answers is not in the document]
Comment (zixi0825) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ProcessDefinitionController.java]: problem fixed

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ProcessDefinitionController.java [hunk 3]
```diff
@@ -95,27 +95,33 @@ public Result createProcessDefinition(@ApiIgnore @RequestAttribute(value = Const
...
+
+
+        logger.info("batch {} process definition, login user:{}, project name:{}, process definition ids:{}，target project name:{}",
+                isCopy?"copy":"move",StringUtils.replaceNRTtoUnderline(loginUser.getUserName()), StringUtils.replaceNRTtoUnderline(projectName), StringUtils.replaceNRTtoUnderline(processDefinitionIds),StringUtils.replaceNRTtoUnderline(targetProjectName));
```
[the remark this answers is not in the document]
Comment (zixi0825) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ProcessDefinitionController.java]: Parameters already exist

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/enums/Status.java [hunk 4]
```diff
@@ -176,6 +176,10 @@
     DELETE_WORKER_GROUP_FAIL(10147,"delete worker group fail ", "删除worker分组失败"),
     COPY_PROCESS_DEFINITION_ERROR(10148,"copy process definition error", "复制工作流错误"),
     USER_DISABLED(10149,"The current user is disabled", "当前用户已停用"),
+    QUERY_USER_CREATED_PROJECT_ERROR(10151,"query user created project error error", "查询用户创建的项目错误"),
+    PROCESS_DEFINITION_IDS_IS_EMPTY(10152,"process definition ids is empty", "工作流IDS不能为空"),
+    COPY_OR_MOVE_PROCESS_DEFINITION_ERROR(10148,"copy or move process definition error", "复制或者移动工作流错误"),
+    MOVE_PROCESS_DEFINITION_ERROR(10150,"move process definition error", "移动工作流错误"),
```
[the remark this answers is not in the document]
Comment (zixi0825) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/enums/Status.java]: problem fixed

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java [hunk 5]
```diff
@@ -295,36 +295,188 @@ private String getResourceIds(ProcessData processData) {
 
     /**
      * copy process definition
-     *
      * @param loginUser   login user
-     * @param projectName project name
      * @param processId   process definition id
      * @return copy result code
      */
```
[the remark this answers is not in the document]
Comment (zixi0825) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java:302]: problem fixed

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java [hunk 6]
```diff
@@ -295,36 +295,188 @@ private String getResourceIds(ProcessData processData) {
...
+                                                     String targetProjectName) throws JsonProcessingException {
 
         Map<String, Object> result = new HashMap<>(5);
+
+        ProcessDefinition processDefinition = processDefineMapper.selectById(processId);
+        if (processDefinition == null) {
+            putMsg(result, Status.PROCESS_DEFINE_NOT_EXIST, processId);
+            return result;
+        } else {
+            Project targetProject = projectMapper.queryByName(targetProjectName);
```
[the remark this answers is not in the document]
Comment (zixi0825) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java]: problem fixed

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java (hunk 1 above)
Comment (zixi0825) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java]: already add project name

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProjectService.java [hunk 7]
```diff
@@ -43,7 +43,7 @@
  *HttpTask./
```
Comment (samz406) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProjectService.java:43]: Modify note description

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ProjectController.java [hunk 8]
```diff
@@ -226,6 +227,23 @@ public Result queryAuthorizedProject(@ApiIgnore @RequestAttribute(value = Consta
...
+     * @return projects which the user create
+     */
+    @ApiOperation(value = "queryProjectCreatedByUser", notes = "QUERY_USER_CREATED_PROJECT_NOTES")
+
+    @GetMapping(value = "/login-user-created-project")
+    @ResponseStatus(HttpStatus.OK)
+    @ApiException(QUERY_USER_CREATED_PROJECT_ERROR)
+    public Result queryProjectCreatedByUser(@ApiIgnore @RequestAttribute(value = Constants.SESSION_USER) User loginUser) {
+        logger.info("login user {}, query authorized project by user id: {}.", StringUtils.replaceNRTtoUnderline(loginUser.getUserName()), StringUtils.replaceNRTtoUnderline(String.valueOf(loginUser.getId())));
+        Map<String, Object> result = projectService.queryProjectCreatedByUser(loginUser, loginUser.getId());
```
Comment (samz406) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ProjectController.java:243]: queryProjectCreatedByUser can define a parameter, you can not userid

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java [hunk 9]
```diff
@@ -295,38 +294,209 @@ private String getResourceIds(ProcessData processData) {
...
+     * @param failedIdList failedIdList
+     * @param processDefinitionIdList processDefinitionIdList
+     */
+    private void doBatchCopyProcessDefinition(User loginUser, Project targetProject, List<String> failedIdList, String[] processDefinitionIdList) {
+        for(String processDefinitionId:processDefinitionIdList){
+            try {
+                Map<String, Object> copyProcessDefinitionResult =
+                        copyProcessDefinition(loginUser,Integer.valueOf(processDefinitionId),targetProject);
+                if (!Status.SUCCESS.equals(copyProcessDefinitionResult.get(Constants.STATUS))) {
+                    failedIdList.add((String) copyProcessDefinitionResult.get(Constants.MSG));
```
[the remark this answers is not in the document]
Comment (zixi0825) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java:413]: In my opinion, simply returning the id does not fully express the reason for the failure, I designed it to return a more complete cause of the error
---------------------------------------------------------------------------------------------------------------------------------
在我看来，仅仅返回id并不能完全表达失败的原因，我这么设计是为了返回更加完整的错误原因

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java [hunk 10]
```diff
@@ -295,38 +294,209 @@ private String getResourceIds(ProcessData processData) {
...
+     * @param processDefinitionIds processDefinitionIds
+     * @param targetProjectId targetProjectId
+     * @param isCopy isCopy
+     * @return
+     */
+    public Map<String, Object> batchCopyOrMoveProcessDefinition(User loginUser,
+                                                          String projectName,
+                                                          String processDefinitionIds,
+                                                          int targetProjectId, boolean isCopy){
+        Map<String, Object> result = new HashMap<>(5);
```
[the remark this answers is not in the document]
Comment (zixi0825) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java:340]: The initial value of HashMap in all Controller and Service in DolphinScheduler is 5, so if you want to modify it, then it is best to unify all changes after discussion

————————————————————————————————————————————-————————
在DolphinScheduler中所有的Controller和Service中HashMap的初始值都为5，所以如果要修改的话，那么最好经过讨论以后，统一全部修改

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ProjectController.java (hunk 8 above)
Comment (zixi0825) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ProjectController.java:243]: The purpose of this function design is to get the project created by the logged-in user, so I think there is no problem using userid

————————————————————————————————————————————————————

这个函数设计的目的就是获取登录用户所创建的项目，所以我认为使用userid没有问题

Comment (samz406) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ProjectController.java:243]: queryProjectCreatedByUser 方法其实传loginUser 值就可以了啊。在方法里面直接loginUser.getId() 获取userid。不影响功能啊。

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (samz406) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java:340]: This was discussed before, you can send an email and change it in a unified way
这个之前讨论过，你可以发个邮件，统一改了

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (samz406) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java:413]: I have two questions:
1 Follow your current design. There may be a splicing return. Part of the error description and part of the id situation are not clear.
2 I think that the id should not be returned here, the process name should be returned, and the front end should be prompted, which of the several process errors is ok. If the id is returned, the front-end operator does not understand which process problem these id refers to

我有两点疑问，
1 按照你现在设计。有可能会出现拼接返回 一部分是错误描述，一部分是id情况 ,这样描述不清晰了。
2 我觉得这里不应该返回id，应该返回流程名称。给前端提示，哪个几个流程错误就ok了，如果返回id，前端操作人员也不理解这些id指的是哪个流程问题。你觉得了

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ProjectController.java (hunk 8 above)
Comment (zixi0825) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ProjectController.java:243]: You're right

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProjectService.java (hunk 7 above)
Comment (zixi0825) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProjectService.java:41]: Makes no sense, I changed it back

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (zixi0825) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java]: I have changed it to the name of the workflow that returned the error. If the error information is returned, it cannot be ruled out that the information is incomplete

我已经将其修改为返回出错的工作流名称了，返回错误描述的话确实不能排除信息不完整的情况

例子：如果工作流不存在的话，返回的信息如下：
从project_test1复制工作流到project_test1错误 : 47[null]

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (Rubik-W) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java]: Hi, This is not fixed.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java (hunk 1 above)
Comment (Rubik-W) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java]: failedIdList -> failedNameList

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (zixi0825) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java:339]: OK!

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (zixi0825) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java]: image
What you see should be old code， the new code has fixed.

**Comment by break60** [comment]
Do not change the front-end .env file

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/BaseService.java [hunk 11]
```diff
@@ -32,7 +32,7 @@
 /**
  * base service
  */
-public class BaseService {
+public class BaseService{
```
Comment (yangyichao-mango) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/BaseService.java:35]: Hi,
Please revert this blank.

Comment (yangyichao-mango) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/BaseService.java:35]: Hi,
Please revert this blank.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java [hunk 12]
```diff
@@ -1256,7 +1427,7 @@ public void importSubProcess(User loginUser, Project targetProject, ArrayNode js
      * @return tree view json data
      * @throws Exception exception
      */
-    public Map<String, Object> viewTree(Integer processId, Integer limit) throws Exception {
+    public Map<String, Object> viewTree(Integer processId, Integer limit) throws Exception{
```
Comment (yangyichao-mango) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java:1430]: Please revert this changes

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ProcessDefinitionController.java [hunk 13]
```diff
@@ -95,27 +95,39 @@ public Result createProcessDefinition(@ApiIgnore @RequestAttribute(value = Const
...
-    @ApiOperation(value = "copyProcessDefinition", notes= "COPY_PROCESS_DEFINITION_NOTES")
+    @ApiOperation(value = "copyOrMoveProcessDefinition", notes= "COPY_OR_MOVE_PROCESS_DEFINITION_NOTES")
     @ApiImplicitParams({
-            @ApiImplicitParam(name = "processId", value = "PROCESS_DEFINITION_ID", required = true, dataType = "Int", example = "100")
+            @ApiImplicitParam(name = "processDefinitionIds", value = "PROCESS_DEFINITION_IDS", required = true, dataType = "String", example = "3,4"),
+            @ApiImplicitParam(name = "targetProjectId", value = "TARGET_PROJECT_ID", required = true, type = "Integer"),
+            @ApiImplicitParam(name = "isCopy", value = "IS_COPY", required = true, type = "boolean")
     })
-    @PostMapping(value = "/copy")
+    @PostMapping(value = "/copy-or-move")
```
Comment (yangyichao-mango) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ProcessDefinitionController.java:113]: Hi,
Will it be better to split copy and move to two api for single responsibility?

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ProcessDefinitionController.java [hunk 14]
```diff
@@ -231,8 +243,8 @@ public Result releaseProcessDefinition(@ApiIgnore @RequestAttribute(value = Cons
     @ResponseStatus(HttpStatus.OK)
     @ApiException(QUERY_DATAIL_OF_PROCESS_DEFINITION_ERROR)
     public Result queryProcessDefinitionById(@ApiIgnore @RequestAttribute(value = Constants.SESSION_USER) User loginUser,
-                                              @ApiParam(name = "projectName", value = "PROJECT_NAME", required = true) @PathVariable String projectName,
-                                              @RequestParam("processId") Integer processId
+                                             @ApiParam(name = "projectName", value = "PROJECT_NAME", required = true) @PathVariable String projectName,
+                                             @RequestParam("processId") Integer processId
```
Comment (yangyichao-mango) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ProcessDefinitionController.java:247]: Please revert the changes about blank, thx a lot, it is suggested to add the blank according to the checkstyle.xml, you can configure the checkstyle.xml[1] to your ide.

[1] https://github.com/apache/incubator-dolphinscheduler/blob/dev/style/checkstyle.xml

**Comment by gabrywu** [comment]
Hi , @samz406 @yangyichao-mango , could this PR be merged?

**Comment by yangyichao-mango** [comment]
> Hi , @samz406 @yangyichao-mango , could this PR be merged?

I think split copy and move to two api will be better, but at present it LGTM, we can improve it later.

**Comment by zixi0825** [comment]
> > Hi , @samz406 @yangyichao-mango , could this PR be merged?
> 
> I think split copy and move to two api will be better, but at present it LGTM, we can improve it later.

i will split it next week

**Comment by gabrywu** [comment]
> > > Hi , @samz406 @yangyichao-mango , could this PR be merged?
> > 
> > 
> > I think split copy and move to two api will be better, but at present it LGTM, we can improve it later.
> 
> i will split it next week

@yangyichao-mango @zixi0825 split in this PR?

**Comment by zixi0825** [comment]
> > > > Hi , @samz406 @yangyichao-mango , could this PR be merged?
> > > 
> > > 
> > > I think split copy and move to two api will be better, but at present it LGTM, we can improve it later.
> > 
> > 
> > i will split it next week
> 
> @yangyichao-mango @zixi0825 split in this PR?

yes

**Comment by gabrywu** [comment]
@zixi0825 Please resolve the conflicting files, then PR can be merged
